azure-proto-core/Resources/SystemAssignedIdentity.cs

```diff
@@ -0,0 +1,155 @@
+
+            if (TenantId.Value.CompareTo(other.TenantId.Value) == 1 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 1)
+                return false;
+
+            if (TenantId.Value.CompareTo(other.TenantId.Value) == 0 && PrincipalId.Value.CompareTo(other.PrincipalId.Value) == 0)
+                return true;
+
+            return false;
+        }
+
+        public static SystemAssignedIdentity Deserialize(JsonElement element)
+        {
+            Optional<Guid> principalId = default;
+            Optional<Guid> tenantId = default;
+            bool isPrincipalIdNull = false;
+            bool isTenantIdNull = false;
+            foreach (var property in element.EnumerateObject())
+            {
+                if (property.NameEquals("principalId"))
+                {
+                    if (property.Value.ValueKind == JsonValueKind.Null)
+                    {
+                        isPrincipalIdNull = true;
+                        continue;
+                    }
+
+                    principalId = Guid.Parse(property.Value.GetString());
+                    continue;
+                }
+
+                if (property.NameEquals("tenantId"))
+                {
+                    if (property.Value.ValueKind == JsonValueKind.Null)
+                    {
+                        isTenantIdNull = true;
+                        continue;
+                    }
+
+                    tenantId = Guid.Parse(property.Value.GetString());
+                    continue;
+                }
+            }
+
+            if (isPrincipalIdNull && isTenantIdNull)
+                return null;
+
+            else if ((isPrincipalIdNull && !isTenantIdNull) || (!isPrincipalIdNull && isTenantIdNull))
+                throw new InvalidOperationException("Either TenantId or PrincipalId were null");
+
+            else
+                return new SystemAssignedIdentity(tenantId, principalId);
+        }
+
+        public static void Serialize(Utf8JsonWriter writer, SystemAssignedIdentity systemAssignedIdentity)
+        {
+            writer.WriteStartObject();
+            writer.WritePropertyName("principalId");
+            if (!Optional.IsDefined(systemAssignedIdentity.PrincipalId))
+                writer.WriteStringValue("null");
+
+            else
+                writer.WriteStringValue(systemAssignedIdentity.PrincipalId.ToString());
+
+            writer.WritePropertyName("tenantId");
+            if (!Optional.IsDefined(systemAssignedIdentity.TenantId))
+                writer.WriteStringValue("null");
+
+            else
+                writer.WriteStringValue(systemAssignedIdentity.TenantId.ToString());
+
+            writer.WriteEndObject();
+            writer.Flush();
+        }
+
+        public static bool Equals(SystemAssignedIdentity original, SystemAssignedIdentity other)
+        {
+            if (original == null)
+                return other == null;
+
+            if (other == null)
```

Comment:
line 146 to 150 are redundant they will be executed inside original.Equals(other)